chalk-engine/src/context.rs

```diff
@@ -401,3 +374,9 @@ pub trait AnswerStream<C: Context> {
     fn any_future_answer(&mut self, test: impl FnMut(&C::InferenceNormalizedSubst) -> bool)
         -> bool;
 }
+
+pub trait Canonical<T, C: Context> {
```

Comment:
I was looking at some of the git history since writing this, and it seems like there was a push to reduce the traits in `Context`, so I'm not sure this trait makes since.

Reply:
Yeah, not sure, I did push away from these sorts of traits at some point. I think it was because, for real flexibility, you wind up wanting to thread `&C` arguments around anyway, since some crates may need access to fields.

Reply:
Oh and I think it was because of coherence reasons, too -- this way, you only needed to own the "context" type, but the other types might be from other crates

Reply:
I think I'd prefer to have the methods collected on the `Context` type, but we could introduce "indirection helpers" if desired to make the code read more nicely. Still, I'm not convinced these methods are needed at all.